### DISCUSSION TITLE
Experience Discussion

### DISCUSSION THREAD

**Discussion by WikiWikiWasp** [Games-Gamers/FamBot#8] [Ideas]
### **What should be the sources members can gain experience from?**
So far, I just have mentions of `fam` _(in various forms)_ giving experience. This is only for the single message containing _at least one `fam`_. So a message with 3 `fam`s won't award experience 3 times, etc. The current parsing takes any form of "fam", even if a substring of a word like `famming` or even unrelated words like `famine`. 

> More string parsing can be done to ensure only "fam" related uses are awarded, but I kinda like the absurdity of it picking out "fam"s in random words lol. I'm open to changing if needed though, or if experience gains from these become a problem.

There is also, currently, experience awarded from the use of the ":FAM:" react. This is either from adding the reaction yourself or simply clicking an existing ":FAM:" reaction originally from the bot or any other member.

> Issue #7 is focused on adding limitations to prevent abuse of the above forms of experience gain. Check the issue for progress on that front.

I'm looking into adding experience for starred posts as well _(Issue #1)_. This will help those who don't follow the meme to get exp more reliably.

Some possible additional sources:
- Specific reactions
- Specific FamBot commands to encourage bot use 
> will need to be mindful of spam
- Joining a voice channel during the `f.time` faming hour
- Streaming in a voice channel during the `f.time` faming hour

### **Experience amount allocation?**
So far, I just have the message source give **1exp**, the react give **3exp**, and (planning on having) starred posts give **5exp**. 

Do these amounts sound fair given the source?

> Someone mentioned the **requirement** of having a member ":FAM:" react to be awarded experience. This could potentially help with Issue #7 but will require a rework to the experience rank up calculations since this would lower the rate at which experience is awarded. But I do like it as a potential way to prevent message/react spamming. 

What about the possibility of something giving users "xp boosts" for a limited time? Something like "a starred post gives x2 exp for the day" or something? Too 'gamey'? 

### **How should rank ups be calculated?**
Currently, I have [required experience for a rank up increase exponentially after a successful rank up](https://github.com/Games-Gamers/FamBot/blob/7eb580d1f30bee1528abcfd3d02eb3ff8f3271a2/main.py#L136). This seems to work as an initial method, but might need refining/reworking depending on the above decisions and how the progress bar calculation (Issue #6) is handled. 

We could have a static percentage added to each rank's required experience for simplicity, tbh. I'm not really partial to a specific method. I just prefer the later ranks _(probably 5+)_ to slow down considerably to add value to the higher ranks as a smaller percentage of members will reach it. This would also extend the "life" of the ranks by preventing most of the members from reaching the current max rank quickly and having no further progression. 

The existing ranks go from 1-10 and just award a Fam title. We can discuss more in a later Discussion what, if any, additional ranks/titles we'd want to add and/or what further "rewards" each rank comes with. Potentially interacting with our other bots, like adding a Fam emoji to the user with the BotRoss bot, access to the repo, etc.

**Comment by WikiWikiWasp** [comment]
Might look into changing the :FAM: react experience gain from **gaining experience** for the member who added the reacttion, to **giving experience** to the member whose post is getting the reaction. 

This would help with promoting posts, similar to the starboard system, but not saved in a different channel, and will help prevent experience farming/spamming.

**Comment by mulchbutler** [comment]
The rank XP values should probably be hard coded instead of calculated for readability sake. I had to spend a while searching for the values to build this table.
| Rank | XP |
|------|---|
| 1 | 0 |
| 2 | 8 |
| 3 | 27 |
| 4 | 65 |
| 5 | 126 |
| 6 | 217 |
| 7 | 344 |
| 8 | 513 |
| 9 | 730 |
| 10 | 1001 |

**Reply by WikiWikiWasp** [reply]
As in, hardcode those values into the code instead of calculating it? Or hardcode into the code but with different values?

**Reply by mulchbutler** [reply]
Whatever the values end up being, hard code them into the code. That way someone can glance at the code and know that rank 5 requires X exp instead of having to calculate it themselves.
